### PR TITLE
Implement hashmap

### PR DESCRIPTION
**Pull request by Darragh-Grealish**
- [x] HashMap implementation using open-addressing and double hashing
    - [x]  Load factory and capacity doubling
    - [x]  Add double hashing instead of linear probing
    - [x]  Using lazy delete, where we flag nodes, should be fine as not many words will be deleted, easy implementation
- [x]  Plumb the hash function into the project replacing “unordered_map”
- [x]  Write test cases for it

### CODE REVIEW

File: tests/test_hash_map.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    map.insert("thirtytwo", 32);
+    map.insert("thirtythree", 33);
+    map.insert("thirtyfour", 34);
+    map.insert("thirtyfive", 35);
+    map.insert("thirtysix", 36);
+    map.insert("thirtyseven", 37);
+    map.insert("thirtyeight", 38);
+    map.insert("thirtynine", 39);
+    map.insert("forty", 40);
+    map.insert("fortyone", 41);
```
Comment (bxrne) [inline tests/test_hash_map.cpp:95]: Might be a better test to check the initial size that we can verify a rehashing had to happen i.e after 40 and then after 50.

**Review by bxrne — CHANGES_REQUESTED** [review]

File: include/hashmap.hpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
+#pragma once
+#include <new>
```
Comment (bxrne) [inline include/hashmap.hpp:2]: <new> and <string> are not needed here

File: tests/test_hash_map.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+TEST(HashMapTest, EraseElements) {
+    HashMap<string, int> map = HashMap<string, int>();
+    map.insert("one", 1);
+    map.insert("two", 2);
+    map.insert("three", 3);
+
+    map.erase("one");
+    map.erase("three");
+
+    EXPECT_EQ(map.size(), 3); // Lazy deletion, size is not decremented
```
[the remark this answers is not in the document]
Comment (Darragh-Grealish) [reply tests/test_hash_map.cpp:44]: I left comment here to explain why size didn't decrement. Should it be removed or moved over the test??

[the remark this answers is not in the document]
Comment (bxrne) [reply tests/test_hash_map.cpp:44]: I misread initially but should probably go above test like this:
https://github.com/Darragh-Grealish/clouseau/pull/11/files/33da0be84b2f41bc37216f5b56362619bd6a639f#r1788654069

File: tests/test_hash_map.cpp (hunk 1 above)
Comment (Darragh-Grealish) [reply tests/test_hash_map.cpp:105]: Have updated that test.
Initially checks size is 0.
Insert 41 elements.
Check size is 41, which confirms that capacity has been doubled, from 40 to 80.

File: tests/test_hash_map.cpp
[the remark this answers is not in the document]
Comment (Darragh-Grealish) [reply tests/test_hash_map.cpp:1]: Good call ! 
Have added comments over each test.

**Comment by Darragh-Grealish** [comment]
Take ur time reviewing, if you're happy with the pull request, I'll fix up the small conflicts and merge into main.

**Comment by bxrne** [comment]
@Darragh-Grealish Looks good tbh. I'd say add a bit in the `DECISIONS.md` re your implementation then we're good to go i'd say

**Review by bxrne — APPROVED** [review]

**Pull request merged**